[QC-946] Users params run type

Barthelemy opened the pull request.

knopers8 commented:
It seems there are some legit test failures:
```
 7/42 Test  #9: testCheckInterface ......................***Failed    1.63 sec
27/42 Test #33: testCustomParameters ....................***Failed    1.54 sec
31/42 Test #37: testMeanIsAbove .........................***Failed    1.83 sec
```

Barthelemy commented:
yes, indeed, I'll fix it

Barthelemy commented:
I realized that I did not test the custom parameters for the check and aggregator. I put back the PR in WIP for the time being.

knopers8 commented:
still broken.

tomas-herman commented:
Hi @Barthelemy, will this PR enable also setting different values of data sampling in the json files based on the run type? E.g. running our digit task with 0.1% sampling for cosmic/noise runs and 0.001% sampling for pp?

Barthelemy commented:
@tomas-herman no, it won't but that's a good idea. I would put it in a different PR though. Could you create a ticket perhaps ?

tomas-herman commented:
Sure, here it is: https://alice.its.cern.ch/jira/browse/QC-962

knopers8 commented on Framework/include/QualityControl/CustomParameters.h:
```diff
@@ -0,0 +1,222 @@
+   * @param defaultValue
+   * @return the value for the given key, runType and beamType or defaultValue if none is found.
+   */
+  std::string atOrDefaultValue(const std::string& key, const std::string& runType, const std::string& beamType, std::string defaultValue) const
+  {
+    try {
+      return mCustomParameters.at(runType).at(beamType).at(key);
+    } catch (const std::out_of_range& exc) {
+      return defaultValue;
+    }
+  }
+
+  /**
+   * Return the value for the given key. If it does not exist, returns the provided default value.
+   * @param key
+   * @param defaultValue
+   * @return the value for the given key. If it does not exist, returns the provided default value.
+   */
+  std::string atOrDefaultValue(const std::string& key, std::string defaultValue) const
+  {
+    try {
+      return mCustomParameters.at("default").at("default").at(key);
+    } catch (const std::out_of_range& exc) {
+      return defaultValue;
+    }
+  }
```
Maybe these two can be joined together by:
```
std::string atOrDefaultValue(const std::string& key, std::string defaultValue = "", const std::string& runType = "default", const std::string& beamType = "default")
```

knopers8 replied:
Then you could consider joining it also with `at` if you used `optional`.

knopers8 commented on Framework/src/InfrastructureSpecReader.cxx:
```diff
@@ -107,9 +107,18 @@ TaskSpec InfrastructureSpecReader::readSpecEntry<TaskSpec>(const std::string& ta
+      }
+    }
+  }
   if (taskTree.count("taskParameters") > 0) {
```
maybe we should complain if users use both at the same time?

Barthelemy replied:
yes, I have put a warning and I actually only use `extendedTaskParameters` in such case.

knopers8 commented on doc/Advanced.md:
```diff
@@ -992,8 +992,77 @@ The same approach can be applied to other actors in the QC framework, like Check
 
 ## Definition and access of task-specific configuration
 
+A task can access custom parameters declared in the configuration file at `qc.tasks.<task_id>.extendTaskParameters` or `qc.tasks.<task_id>.taskParameters`. They are stored inside an object of type `CustomParameters` named `mCustomParameters`, which is a protected member of `TaskInterface`.
```
```suggestion
A task can access custom parameters declared in the configuration file at `qc.tasks.<task_id>.extendedTaskParameters` or `qc.tasks.<task_id>.taskParameters`. They are stored inside an object of type `CustomParameters` named `mCustomParameters`, which is a protected member of `TaskInterface`.
```

knopers8 commented on doc/Advanced.md:
```diff
@@ -992,8 +992,77 @@ The same approach can be applied to other actors in the QC framework, like Check
+            "pp": {
+              "myOwnKey1": "myOwnValue1c"
+            },
+            "PbPb": {
```
Are "pp" and "PbPb" the correct examples of the strings to be used? The same comment for "physics" and "cosmics".

Barthelemy replied:
I have added an example for the Run Type that is defined in O2 in "DataFormatsParameters/ECSDataAdapters.h". 
Concerning the beam type I am still waiting for the information.

Barthelemy replied:
ok, found it, added doc.

knopers8 reviewed (changes requested):
Thanks, please see my comments.

One thing that I don't understand is how the string for run type can be compared to `Activity.type` which is int. Are the users supposed to use numbers actually?

knopers8 commented on doc/Advanced.md:
```diff
@@ -992,8 +992,77 @@ The same approach can be applied to other actors in the QC framework, like Check
 
 ## Definition and access of task-specific configuration
```
Please update the index.

Barthelemy commented on Framework/include/QualityControl/CustomParameters.h:
```diff
@@ -0,0 +1,222 @@
+
+#include <string>
+#include <unordered_map>
+#include <iostream>
```
indeed, done.

Barthelemy commented:
I agree that it is not extensible and this is to address the request from RC concerning the Run type and the beam type. 

I had to maintain the interface of CustomParameters similar to `vector` to avoid updating hundreds of files. 

The beam type is still not defined, ie. I don't have a translation between the int we get from BK and the string. Once it is there I will update the example and the doc.

Barthelemy commented:
@knopers8 I have addressed your comments as best as I could. Beam type is defined in the doc. Run type is defined in a class in O2. I have added a link to the doc. 
I also added an example in SkeletonTask.

knopers8 commented on doc/Advanced.md:
````diff
@@ -992,8 +997,80 @@ The same approach can be applied to other actors in the QC framework, like Check
+```
+It allows to have variations of the parameters depending on the run and beam types. The proper run types can be found here: https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Parameters/include/DataFormatsParameters/ECSDataAdapters.h#L54. The `default` can be used 
+to ignore the run or the beam type. 
+The beam type is one of the following: `PROTON-PROTON`, `Pb-Pb`, `Pb-PROTON`
````
Can you use these strings in the examples as well?

Barthelemy replied:
I found one instance that I  changed

knopers8 commented on Framework/include/QualityControl/CustomParameters.h:
```diff
@@ -0,0 +1,222 @@
+    if (mCustomParameters.count(runType) > 0 && mCustomParameters[runType].count(beamType) > 0) {
+      return mCustomParameters[runType][beamType];
+    }
+    throw std::out_of_range("Unknown beam or run: " + runType + ", " + beamType);
```
this is not resolved yet

knopers8 commented on Framework/include/QualityControl/CustomParameters.h:
```diff
@@ -0,0 +1,213 @@
+ *   cout << "value for key `key` : " << cp.at("key", "physics", "pp") << endl;
+ *
+ */
+class CustomParameters
```
I don't quite understand why we should recompile these methods for each time someone includes this header. Perhaps everything could be moved to a source file?

Barthelemy replied:
ok I had missed your previous comment
fixed.
